Fixed bugs

gi-dellav opened the pull request:
Project wouldn't compile because of dependency conflicts and missing modules in `gbam_tools/lib.rs`.


---


Tested on Debian Stable.

gi-dellav committed:
Fixed deps

gi-dellav committed:
Fixed missing modules

gi-dellav commented:
But the main test is still now working, it get stuck at 10.1 MB

❯ time ./target/release/gbam_binary -c test_data/little.bam -o test.gbam                                                                          ⠁ [00:00:00] [#####################>----------------------------------------------------------------------------------] 10.10MiB/49.21MiB (39.74MiB/s, 1s)

{here it was manually interrupted)
________________________________________________________
Executed in   12.14 secs    fish           external
   usr time   44.74 secs  541.00 micros   44.74 secs
   sys time    0.22 secs  106.00 micros    0.22 secs

gi-dellav commented:
<img width="1920" height="1080" alt="immagine" src="https://github.com/user-attachments/assets/2365936b-8b5b-40a1-8faf-8031105a6366" />

The perf is looking normal, most time spent on Brotli functions (as expected)

AndreaGuarracino commented:
`pytest`

<img width="3254" height="582" alt="image" src="https://github.com/user-attachments/assets/64e427c4-ecfb-4a24-8fa0-4d5706b51a7f" />

AndreaGuarracino commented:
@gi-dellav, some random subsets of the `little.bam` file:

[little.s01.zip](https://github.com/user-attachments/files/21196339/little.s01.zip)
[little.s02.zip](https://github.com/user-attachments/files/21196345/little.s02.zip)

gi-dellav commented:
Other note: Running the tool with `--thread-num 1` still spawns 4 threads allocated to 4 different CPU cores, so the program ignores the given parameter

gi-dellav commented:
@AndreaGuarracino @NickRoz1 

Tested the two subset, the results aren't improved, I don't think that the BAM file is the issue:

(Note: the logs are produced using some simple delta timers that i wrote inside of the relevant function to measure spent time, as `perf` wasn't being useful)

## Timed log of `little.s01.bam`
```
get_compr_block in 532ns
Updated meta in  71ns
Block prepared (not including block_info) in 21.916µs
-----------
Block compressed in 38.422µs
get_compr_block in 764ns
Updated meta in  51ns
Block prepared (not including block_info) in 11.667µs
-----------
Block compressed in 53.815µs
get_compr_block in 262ns
Updated meta in  58ns
Block prepared (not including block_info) in 3.026µs
-----------
Block compressed in 16.828µs
get_compr_block in 128ns
Updated meta in  54ns
Block prepared (not including block_info) in 2.453µs
-----------
Block compressed in 31.673µs
get_compr_block in 218ns
Updated meta in  57ns
Block prepared (not including block_info) in 2.418µs
-----------
Block compressed in 12.47µs
get_compr_block in 134ns
Updated meta in  49ns
Block prepared (not including block_info) in 5.239µs
-----------
Block compressed in 566.743µs
get_compr_block in 1.774µs
Updated meta in  46ns
Block prepared (not including block_info) in 4.976µs
-----------
Block compressed in 30.005µs
get_compr_block in 122ns
Updated meta in  44ns
Block prepared (not including block_info) in 15.422µs
-----------
Block compressed in 901.27µs
get_compr_block in 429ns
Block position seeked in 1.92µs
Generated FileMeta in 3.48µs
Written to buffer in 268.225µs
Updated meta in  276.216µs
Block prepared (not including block_info) in 279.726µs
-----------
Block compressed in 25.081µs
get_compr_block in 165.454584ms
Block position seeked in 1.637µs
Generated FileMeta in 5.716µs
Written to buffer in 1.142µs
Updated meta in  10.218µs
Block prepared (not including block_info) in 165.482479ms
-----------
Block compressed in 2.622325ms
get_compr_block in 34.758676ms
Block position seeked in 16.349µs
Generated FileMeta in 140.707µs
Written to buffer in 1.944µs
Updated meta in  209.424µs
Block prepared (not including block_info) in 35.046107ms
-----------
Block compressed in 33.204µs
get_compr_block in 44.546186ms
Block position seeked in 13.833µs
Generated FileMeta in 44.622µs
Written to buffer in 1.281µs
Updated meta in  100.483µs
Block prepared (not including block_info) in 44.859139ms
-----------
Block compressed in 17.354µs
get_compr_block in 1.332µs
Block position seeked in 6.233µs
Generated FileMeta in 36.272µs
Written to buffer in 561ns
Updated meta in  92.73µs
Block prepared (not including block_info) in 152.1µs
-----------
Block compressed in 1.028471ms
get_compr_block in 455.200295ms
Block position seeked in 21.315µs
Generated FileMeta in 24.986µs
Written to buffer in 1.962µs
Updated meta in  31.534µs
Block prepared (not including block_info) in 455.246087ms
-----------
Block compressed in 4.668316ms
get_compr_block in 1.418944902s
Block position seeked in 31.187µs
Generated FileMeta in 35.916µs
Written to buffer in 12.937µs
Updated meta in  53.692µs
Block prepared (not including block_info) in 1.419016174s
-----------
Block compressed in 12.956553ms
get_compr_block in 1.945215416s
Block position seeked in 1.535µs
Generated FileMeta in 32.487µs
Written to buffer in 47.732µs
Updated meta in  137.832µs
Block prepared (not including block_info) in 1.945786772s
-----------
Block compressed in 23.73µs
get_compr_block in 1.618624195s
Block position seeked in 1.885µs
Generated FileMeta in 32.404µs
Written to buffer in 169.704µs
Updated meta in  266.702µs
Block prepared (not including block_info) in 1.619199797s
-----------
Block compressed in 3.604686ms
get_compr_block in 1.431728619s
Block position seeked in 2.098µs
Generated FileMeta in 34.02µs
Written to buffer in 32.288µs
Updated meta in  125.381µs
Block prepared (not including block_info) in 1.432093965s
-----------
Block compressed in 17.924µs
get_compr_block in 988.728974ms
Block position seeked in 2.736µs
Generated FileMeta in 30.895µs
Written to buffer in 2.83µs
Updated meta in  84.456µs
Block prepared (not including block_info) in 989.073594ms
-----------
Block compressed in 1.869991ms
Block position seeked in 32.999µs
Generated FileMeta in 41.417µs
Written to buffer in 75.938µs
Block position seeked in 830ns
Generated FileMeta in 4.261µs
Written to buffer in 107.917µs
Block position seeked in 307ns
Generated FileMeta in 1.935µs
Written to buffer in 111.323µs
Block position seeked in 388ns
Generated FileMeta in 2.19µs
Written to buffer in 133.877µs
Block position seeked in 326ns
Generated FileMeta in 1.949µs
Written to buffer in 100.033µs
Block position seeked in 292ns
Generated FileMeta in 1.837µs
Written to buffer in 124.18µs
Block position seeked in 292ns
Generated FileMeta in 1.527µs
Written to buffer in 459.201µs
Block position seeked in 434ns
Generated FileMeta in 2.138µs
Written to buffer in 421.336µs

________________________________________________________
Executed in   42.43 secs    fish           external
   usr time  133.11 secs  819.00 micros  133.11 secs
   sys time    0.53 secs    0.00 micros    0.53 secs
```

## Timed log of `little.s02.bam`

```
get_compr_block in 563ns
Updated meta in  49ns
Block prepared (not including block_info) in 29.821µs
-----------
Block compressed in 37.907µs
get_compr_block in 479ns
Updated meta in  63ns
Block prepared (not including block_info) in 11.213µs
-----------
Block compressed in 1.394463ms
get_compr_block in 545ns
Updated meta in  87ns
Block prepared (not including block_info) in 41.408µs
-----------
Block compressed in 3.06728ms
get_compr_block in 559ns
Updated meta in  63ns
Block prepared (not including block_info) in 15.037µs
-----------
Block compressed in 747.028µs
get_compr_block in 574ns
Updated meta in  76ns
Block prepared (not including block_info) in 14.382µs
-----------
Block compressed in 3.144638ms
get_compr_block in 1.028µs
Updated meta in  60ns
Block prepared (not including block_info) in 6.392µs
-----------
Block compressed in 3.147096ms
get_compr_block in 581ns
Updated meta in  52ns
Block prepared (not including block_info) in 3.527µs
-----------
Block compressed in 4.827808ms
get_compr_block in 9.422µs
Updated meta in  192ns
Block prepared (not including block_info) in 43.926µs
-----------
Block compressed in 52.99µs
get_compr_block in 350.186863ms
Block position seeked in 1.71µs
Generated FileMeta in 4.385µs
Written to buffer in 1.403µs
Updated meta in  32.024µs
Block prepared (not including block_info) in 350.244401ms
-----------
Block compressed in 5.352563ms
get_compr_block in 82.86µs
Block position seeked in 51.047µs
Generated FileMeta in 63.766µs
Written to buffer in 501.213µs
Updated meta in  643.59µs
Block prepared (not including block_info) in 749.503µs
-----------
Block compressed in 2.339752ms
get_compr_block in 371.27887ms
Block position seeked in 2.03µs
Generated FileMeta in 6.446µs
Written to buffer in 1.009µs
Updated meta in  80.066µs
Block prepared (not including block_info) in 371.373782ms
-----------
Block compressed in 1.875569ms
get_compr_block in 251.470092ms
Block position seeked in 15.867µs
Generated FileMeta in 19.811µs
Written to buffer in 872ns
Updated meta in  25.679µs
Block prepared (not including block_info) in 251.510431ms
-----------
Block compressed in 4.472715ms
get_compr_block in 333.999884ms
Block position seeked in 16.783µs
Generated FileMeta in 224.334µs
Written to buffer in 1.804µs
Updated meta in  463.136µs
Block prepared (not including block_info) in 334.547289ms
-----------
Block compressed in 40.971µs
get_compr_block in 1.33355739s
Block position seeked in 217.375µs
Generated FileMeta in 258.291µs
Written to buffer in 15.845µs
Updated meta in  334.162µs
Block prepared (not including block_info) in 1.334051514s
-----------
Block compressed in 11.245959ms
get_compr_block in 2.379881415s
Block position seeked in 2.436µs
Generated FileMeta in 32.641µs
Written to buffer in 42.874µs
Updated meta in  135.911µs
Block prepared (not including block_info) in 2.380137709s
-----------
Block compressed in 15.69206ms
get_compr_block in 4.649767575s
Block position seeked in 3.776µs
Generated FileMeta in 12.467µs
Written to buffer in 250.528µs
Updated meta in  295.856µs
Block prepared (not including block_info) in 4.650090722s
-----------
Block compressed in 24.989877ms
get_compr_block in 2.795436621s
Block position seeked in 2.088µs
Generated FileMeta in 11.732µs
Written to buffer in 423.534µs
Updated meta in  454.394µs
Block prepared (not including block_info) in 2.795935815s
-----------
Block compressed in 31.972µs
get_compr_block in 4.041356664s
Block position seeked in 2.319µs
Generated FileMeta in 11.566µs
Written to buffer in 105.985µs
Updated meta in  170.582µs
Block prepared (not including block_info) in 4.041587164s
-----------
Block compressed in 6.89031ms
get_compr_block in 8.456674047s
Block position seeked in 3.391µs
Generated FileMeta in 75.206µs
Written to buffer in 292.671µs
Updated meta in  450.094µs
Block prepared (not including block_info) in 8.4580515s
-----------
Block compressed in 32.484µs
get_compr_block in 31.977742742s
Block position seeked in 2.417µs
Generated FileMeta in 20.865µs
Written to buffer in 685.732µs
Updated meta in  782.444µs
Block prepared (not including block_info) in 31.978576263s
-----------
Block compressed in 20.328µs
get_compr_block in 3.977602794s
Block position seeked in 2.402µs
Generated FileMeta in 39.955µs
Written to buffer in 511.265µs
Updated meta in  627.565µs
Block prepared (not including block_info) in 3.978955642s
-----------
Block compressed in 24.354464ms
get_compr_block in 2.010063881s
Block position seeked in 2.259µs
Generated FileMeta in 6.989µs
Written to buffer in 35.288µs
Updated meta in  65.563µs
Block prepared (not including block_info) in 2.010156037s
-----------
Block compressed in 13.502772ms
Block position seeked in 2.248µs
Generated FileMeta in 14.765µs
Written to buffer in 196.283µs
Block position seeked in 1.521µs
Generated FileMeta in 8.079µs
Written to buffer in 228.524µs
Block position seeked in 1.965µs
Generated FileMeta in 7.748µs
Written to buffer in 357.586µs
Block position seeked in 3.425µs
Generated FileMeta in 12.085µs
Written to buffer in 528.227µs
Block position seeked in 1.877µs
Generated FileMeta in 5.759µs
Written to buffer in 561.414µs
Block position seeked in 681ns
Generated FileMeta in 3.144µs
Written to buffer in 195.093µs
Block position seeked in 1.078µs
Generated FileMeta in 21.537µs
Written to buffer in 716.638µs
Block position seeked in 1.226µs
Generated FileMeta in 3.557µs
Written to buffer in 850.005µs

________________________________________________________
Executed in   85.88 secs    fish           external
   usr time  274.26 secs  964.00 micros  274.26 secs
   sys time    0.84 secs    0.00 micros    0.84 secs
```

## Final results

Time of compressions:

- `little.s01.bam`: ~42sec
- `little.s02.bam`:  ~85sec

Size of compressions:

- `little.s01.sbam`: 5.6MB => 5.3MB 
- `little.s02.sbam`: 11MB => 10MB

Deflation:

- `little.s01.sbam`: ~5.35% compression
- `little.s02.sbam`: ~9% compression

REALLY IMPORTANT NOTE: The delta between the different calls of `flush_field_buffer` are still crazy, with some blocks prepared in <1ms and some needed 60 seconds, considering that block size should be static.

The performance is still... depressing.

--

[SBAM Output Files](https://we.tl/t-XG622PGbri) (Sorry for WeTransfer, Github doesn't want to work for some reasons with my files) 

P.S. I just compressed the SBAM files to ZIP because of WeTransfer and... what... why is the ZIP compression deflating more than the SBAM algorithm? I don't know, but it could be interesting to add as a quick final step to the compression pipeline, as ZIP is a really fast both for compression and decompression. 

I'll leave here the observation data about the ZIP compression:

Size of compressions:
- `little.s01.sbam`: 5.3MB => 4.6MB
- `little.s02.sbam`: 10MB => 8.6MB

Deflations:
- `little.s01.sbam`: 13% compression
- `little.s02.sbam`: 14% compression

The same phenomena doesn't happen if i try to ZIP the BAM file (0% deflation, as expected).

gi-dellav commented:
So, i fixed up a bit my internal timers, the performance issues are derived from `get_compr_block` in `compressor.rs`, which sounds really weird, as it should just drain the rx buffer and get the relevant `CompressTask`.

I worked with `flush_field_buffer` and my workaround was to call `get_compr_block` after the compression and use an intermediate buffer in order to process data, the performance improvement is there but it isn't anything crazy, but now the log shows (as it should, and as perf says) that most of the time is spent in the `compress_block` function in `compressor.rs`, but I will later try to dissect this function more, as I don't that is reasonable to spend 1 minute with 4 CPU cores just to compress 5MB.

I will now make a PR with the updated, slighlty faster, version of `flush_field_buffer`

Also, I tried to check the last commits to search for performance regressions (from what I understand, the performance at some point in the past was better), and i tried `7f352e1`, `bb5ee27`, `157cc86`, `49f3e53` (which also breaks because some empty files were removed, but i already fixed it in my original pull request) and all of those are broken both because of the dependencies issues and because of the missing files (check my first PR), so I wanted to ask if they were compilable at the time on your machine and if yes, how.

gi-dellav commented:
I understand that the performance issue now is just Brotli: I see that it's configured at 11 as depth level... isn't it too much (considering that it's the maximum compression level supported by Brotli)?   

I ran some experiments with 4, 5, 6, 7, 8 and 11 as levels, check them for yourself (obviously, 11 will always give the most compressed file, but the time are just excessive in my opinion, at least we should give the users the ability to choose).

### Time
- Level 4: ~5 sec
- Level 5: ~12 sec
- Level 6: ~15 sec
- Level 7: ~27 sec
- Level 8: ~37 sec
- Level 11: ~403 sec

### Compression Rate

(NOTE: The third number derives from running ZIP as an additional step in the pipeline, consider that in all cases it takes less than 1 second on my machine, and compare it to the other results...)

- Level 4: 50MB => 61 MB => 55MB
- Level 5: 50MB => 50MB => 43MB
- Level 6: 50MB => 48MB => 40MB
- Level 7: 50MB => 46MB => 39MB
- Level 8: 50MB => 46MB => 39MB (Same as Level 7, but slower...)
- Level 11: 50MB => 42MB => 36MB (I was expecting more from you, Level 11...)

### Compression Speed Rate 

I decided to include this useful measurement that i made `(original_size-compressed_size)/(original_size)/(compression_time)` in order to have an idea for how much the compression rate grows based on the time needed to run the compression itself.

- Level 5: ~1.16% deflation per second
- Level 6: ~0.93% deflation per second
- Level 7: ~0.81% deflation per second
- Level 8: ~0.59% deflation per second
- Level 11: ~0.07% deflation per second

### Additional logs

From `du --human-readable test_data/little.bam test.brotli*.gbam test.brotli*.gbam.zip` with all relevant files:

```
50M	test_data/little.bam
61M	test.brotli4.gbam
50M	test.brotli5.gbam
48M	test.brotli6.gbam
46M	test.brotli7.gbam
46M	test.brotli8.gbam
42M	test.brotli11.gbam
55M	test.brotli4.gbam.zip
43M	test.brotli5.gbam.zip
40M	test.brotli6.gbam.zip
39M	test.brotli7.gbam.zip
39M	test.brotli8.gbam.zip
36M	test.brotli11.gbam.zip
```

### Conclusions

We need to find the balance of parameters for Brotli; I think that something like Level 6 or 7 might be enough.

Also, interesting observation, the more the compression level grows, not only the compute time grows (as expected), but the deflation/speed rate decreases, indicating that the time needed grows faster than the advantage given by the harder compression level.

chris7716 commented:
I think the reason why you see more compression after zipping is that, the header and the block metadata would be compressed by zipping. When flushing the column / column block, we only compress the particular column block.

NickRoz1 commented:
@gi-dellav 
would you consider also looking into C implementation? there is way less clutter and were keeping it single threaded for the meantime
easier to use libraries also..
still the compression levels are far from CRAM. the question we have is whether the compression rate of CRAM arises due to the complexity of the format, or just a handful of of powerful small techniques. paper is not trivial. if the latter is the case we could easily approach cram compression levels by plugging the techniques into ours c code. 

perhaps examining htslib cram writing routines would be the easiest way to go. there must be a way to trace an execution graph of C code. that way one could study what happens when your rub samtools bam to cram and see if there are specific functions/compression methods that do most of the heavy lifting

chris7716 commented:
One more point is that brotli is not the best for all columns. I have implemented codec support per column here https://github.com/NickRoz1/gbam/pull/20/files#diff-fcf051526facc086ca3d400896194c0532e836f9f72fba9a4815d7a8f916c7c5

Based on my previous analysis, I have found that Zstd performs better in `RawCigar`, `RawQual` and `RawTags` columns. That combination brings 50MB file down to 40MB.
<img width="508" height="422" alt="image" src="https://github.com/user-attachments/assets/62675bfd-66dd-4671-ab20-be13de094c64" />

gi-dellav commented:
@chris7716 I don't think that the observation about metadata can be applied to the fullest to this case, as zip is on all levels of brotli always doing around 10-15% of deflation after the GBAM execution, this would mean that the metadata inside of the file is more than 15% (if we assume 50% compression ratio on metadata, it would mean that 30% of the GBAM is metadata) if only the metadata is compressed.

A good idea could be to either add an experimental flag to skip the writing of the metadata (obviously not to be used for decoding, just to calculate the amount of metadata inside the file).

gi-dellav commented:
@NickRoz1 I will later give a look at the C development branch, but I have a great amount of experience in Rust development, while I didn't study (yet) the C programming language at a good enough level.

Also, during the hackaton I'm in, my project is to optimize the alghorithm parameters (both the compressions alghorithms and levels used for each BAM column and for metadata) and I'm going to implement the tool for evolutionary research for compression optimization in Rust as I never wrote full projects in C.

gi-dellav commented:
@chris7716 Can you push the per-column support to master or to another branch in the repo? I'd like to take your already developed solution in order to have a better base on which to build the evolutionary optimization solver.

gi-dellav commented:
If you check #24, I implemented ZX and chose better compression levels in order to speed up processing with every type of compression.

@chris7716 I will now work on metadata compression and I will merge on my local branch your per-column compression implementation.

gi-dellav commented:
I think that I will rewrite the analysis tool implemented in Python in #20 as:

1. The code will be better integrated with the rest of the codebase (so written in Rust)
2. The code needs to be adapted to metadata compression
3. The code needs to be re-tested with compression levels that rappresent compression/time ratios (the last commits used excessive compression levels)
